Build fails when building from a git worktree: submodule git metadata is outside the container bind-mount (breaks scapy version derivation)

## Description

Building from a **`git worktree`** of `sonic-buildimage` (instead of a full standalone clone) fails because the build containers only bind-mount the worktree directory, not the parent repository's `.git` directory where worktree/submodule git metadata actually lives. As a result, any in-container `git` command run against a submodule fails with `fatal: not a git repository`, which silently breaks version-string derivation for at least the `scapy` python wheel and aborts the build.

There are valid use cases for worktrees (e.g. building multiple branches/commits from a single bare-ish clone without re-downloading ~GBs of submodules per branch), so it would be nice for the build system to support, or at least clearly document/detect, worktree mode.

## Steps to reproduce

```bash
# Existing full clone with submodules
cd /data/sonic-buildimage          # origin = sonic-net/sonic-buildimage

# Create a worktree for master and build the VS image from it
git worktree add /data/sonic-buildimage-master -b master-latest
cd /data/sonic-buildimage-master
git fetch origin master && git reset --hard origin/master
git submodule update --init --recursive

make configure PLATFORM=vs
make SONIC_BUILD_JOBS=4 target/sonic-vs.img.gz
```

## Observed result

Build fails at the `scapy` wheel step:

```
Created wheel for scapy: filename=scapy-2026.6.20-py3-none-any.whl ...
Building Wheels package target/python-wheels/bookworm/scapy-2.6.1.dev0-py3-none-any.whl
mv: cannot stat 'src/scapy/dist/scapy-2.6.1.dev0-py3-none-any.whl': No such file or directory
[  FAIL LOG END  ] [ target/python-wheels/bookworm/scapy-2.6.1.dev0-py3-none-any.whl ]
make: *** [slave.mk:1081: target/python-wheels/bookworm/scapy-2.6.1.dev0-py3-none-any.whl] Error 1
```

The whole build is also spammed with thousands of:

```
fatal: not a git repository: /data/sonic-buildimage/.git/worktrees/sonic-buildimage-master/...
```

## Root cause

In a worktree, a submodule's `.git` is a **pointer file**, not a directory. For example `src/scapy/.git` contains:

```
gitdir: ../../../sonic-buildimage/.git/worktrees/sonic-buildimage-master/modules/src/scapy
```

The build container only bind-mounts the worktree dir into `/sonic`:

```
docker run ... -v /data/sonic-buildimage-master:/sonic -w /sonic ... sonic-slave-bookworm ...
```

So inside the container that gitdir resolves to
`/sonic/src/scapy/../../../sonic-buildimage/.git/worktrees/...`, i.e. it points **outside** the mount, to a path that does not exist in the container. Every in-container `git` call against a submodule therefore returns `rc=128 / fatal: not a git repository`.

`scapy`'s version logic (`scapy/__init__.py::_version()`) tries, in order: `SCAPY_VERSION` env → `scapy/VERSION` file → git-archive subst → **`git describe`** → **date fallback** (`strftime('%Y.%m.%d')`). The only method that yields the pinned `2.6.1.dev0` is `git describe`, and it is wrapped in `except Exception: pass`. With git unreachable in-container it falls through to the date, producing `scapy-2026.6.20-...whl`, while `rules/scapy.mk` expects `scapy-2.6.1.dev0-...whl` → `mv` fails.

This was confirmed with instrumentation inside the in-container build:

```
SONIC-SCAPY-DIAG: enter _version_from_git_describe pkgdir='/sonic/src/scapy/scapy' parent='/sonic/src/scapy' dotgit_exists=True
SONIC-SCAPY-DIAG: git FAILED rc=128 cmd='git describe --tags --always --long' cwd='/sonic/src/scapy/scapy'
   stderr='fatal: not a git repository: /sonic/src/scapy/../../../sonic-buildimage/.git/worktrees/sonic-buildimage-master/modules/src/scapy'
SONIC-SCAPY-DIAG: Method3(git describe) FAIL CalledProcessError: ... exit status 128.
SONIC-SCAPY-DIAG: falling through to DATE fallback (THIS IS THE BUG)
```

Note `dotgit_exists=True` — so the existing patch `src/scapy.patch/0001-Fix-version-string-generation-when-scapy-is-a-submod.patch` (which only removes an `os.path.isdir('.git')` precheck) does **not** help here; the failure is one level deeper, in `git describe` itself.

`scapy` is just the first casualty; any build step that shells out to `git` in a submodule under a worktree is exposed to the same class of failure.

## Expected result

Building from a `git worktree` should either work, or fail fast with a clear message, rather than silently producing wrong version strings deep into the build.

## Possible fixes / mitigations

- **Mount the parent `.git`/common dir** into the build containers when the source tree is a worktree (e.g. also `-v <common-git-dir>:<same-path>`), or rewrite gitdir pointers to in-container paths.
- **Make version derivation robust to unreachable git** — e.g. honor `SCAPY_VERSION` from the build env for the scapy wheel (verified workaround: `SCAPY_VERSION=2.6.1.dev0` makes `_version()` return the pinned version regardless of git), and similarly for any other component that derives versions from git at build time.
- **Detect worktree mode** in `make configure` / `Makefile.work` and emit a clear warning/guard if it is unsupported.

## Workaround

- Build from a regular full `git clone` (self-contained `.git/`) instead of a worktree, **or**
- Export `SCAPY_VERSION=2.6.1.dev0` (matching `rules/scapy.mk`) into the build environment.

## Environment

- Repo: `sonic-net/sonic-buildimage` master @ `be9722abb`
- Source tree: a `git worktree` (`/data/sonic-buildimage-master`, branch `master-latest` tracking `origin/master`) of a full clone at `/data/sonic-buildimage`
- Target: `target/sonic-vs.img.gz` / `target/docker-sonic-vs.gz`, `PLATFORM=vs`, bookworm slave

--- a/Makefile.work
+++ b/Makefile.work
@@ -375,7 +375,7 @@ override DOCKER_BUILDER_WORKDIR := "/$(SONIC_WT_BASENAME)"
 # both on the host and inside the container (the worktree dir and the parent
 # .git keep the same relative relationship in both locations). Run as the first
 # step of the in-container build command.
-override SONIC_WT_GITDIR_FIXUP := echo "gitdir: ../$(SONIC_WT_PARENT_BASENAME)/.git/worktrees/$(SONIC_WT_BASENAME)" > /$(SONIC_WT_BASENAME)/.git;
+override SONIC_WT_GITDIR_FIXUP := echo gitdir: ../$(SONIC_WT_PARENT_BASENAME)/.git/worktrees/$(SONIC_WT_BASENAME) > /$(SONIC_WT_BASENAME)/.git;
 endif
 endif
 endif

--- a/Makefile.work
+++ b/Makefile.work
@@ -319,6 +319,68 @@ ifeq ($(DOCKER_BUILDER_WORKDIR),)
 override DOCKER_BUILDER_WORKDIR := "/sonic"
 endif
 
+# Empty by default; only set when building from a git worktree (see below). For
+# a normal clone it expands to nothing, so the in-container build command is
+# unchanged.
+SONIC_WT_GITDIR_FIXUP :=
+
+# Support building from a git worktree (https://github.com/sonic-net/sonic-buildimage/issues/28002).
+#
+# In a normal clone, $(PWD)/.git is a real directory mounted into the container
+# as part of $(PWD):/sonic, so in-container git (including submodule git used by
+# e.g. src/sonic-frr) works. In a git worktree, $(PWD)/.git is a *file* pointing
+# into the parent repo's git dir, and each submodule's .git pointer plus its
+# core.worktree back-reference resolve to paths OUTSIDE the $(PWD):/sonic mount,
+# so in-container git fails with "fatal: not a git repository" (Error 128).
+#
+# To make worktree builds work, additionally bind-mount, only when building from
+# a worktree (for a normal clone these are empty and nothing changes):
+#   1. the parent repo's common git dir at the same absolute path the worktree's
+#      .git pointers reference inside the container, and
+#   2. the worktree tree itself at its parent-relative basename path, so each
+#      submodule gitdir's core.worktree back-reference resolves onto the real
+#      tree instead of a non-existent path.
+# We also run the build from that basename path (DOCKER_BUILDER_WORKDIR), so the
+# work tree git infers from core.worktree matches the build working directory;
+# otherwise git reports the submodule is not a work tree and tools like stg fail
+# with "Not inside a git worktree". /sonic stays mounted for paths that
+# reference it directly. Both mounts are read-write because package builds such
+# as FRR run git fetch/checkout, stg, and gbp dch, which write to the git dir.
+#
+# Note: these helper variables are deliberately NOT named GIT_DIR / GIT_COMMON_DIR
+# because those are environment variable names git itself honors; reusing them
+# here would leak into child git invocations and break them.
+SONIC_WT_GIT_DIR := $(shell git -C $(PWD) rev-parse --git-dir 2>/dev/null)
+SONIC_WT_GIT_COMMON_DIR := $(shell git -C $(PWD) rev-parse --git-common-dir 2>/dev/null)
+ifeq ($(GIT_WORKTREE_MOUNTS),)
+ifneq ($(SONIC_WT_GIT_DIR),$(SONIC_WT_GIT_COMMON_DIR))
+ifneq ($(SONIC_WT_GIT_COMMON_DIR),)
+# Absolute path to the parent repo's common git dir (e.g. /data/foo/.git).
+SONIC_WT_GIT_COMMON_DIR_ABS := $(shell cd $(PWD) && realpath $(SONIC_WT_GIT_COMMON_DIR) 2>/dev/null)
+# Basename of the parent repo working dir (the dir that contains .git), e.g. foo.
+SONIC_WT_PARENT_BASENAME := $(shell basename $(patsubst %/.git,%,$(SONIC_WT_GIT_COMMON_DIR_ABS)))
+# Basename of this worktree, used by submodule core.worktree back-references.
+SONIC_WT_BASENAME := $(shell basename $(PWD))
+ifneq ($(SONIC_WT_GIT_COMMON_DIR_ABS),)
+override GIT_WORKTREE_MOUNTS := -v $(SONIC_WT_GIT_COMMON_DIR_ABS):/$(SONIC_WT_PARENT_BASENAME)/.git:rw -v $(PWD):/$(SONIC_WT_BASENAME):rw
+# Run the build from the worktree-basename mount so git's inferred work tree
+# (from each submodule's core.worktree back-reference) matches the working dir.
+override DOCKER_BUILDER_WORKDIR := "/$(SONIC_WT_BASENAME)"
+# git worktree add records the top-level worktree .git as an ABSOLUTE gitdir
+# path into the parent repo, unlike submodule .git pointers which are relative
+# (the init target normalizes those). That absolute host path does not exist at
+# the same location inside the container, so in-container git fails and slave.mk
+# derives an empty SONIC_IMAGE_VERSION, which produces an invalid ".0-" docker
+# image tag. Normalize the top-level pointer to a relative gitdir so it resolves
+# both on the host and inside the container (the worktree dir and the parent
+# .git keep the same relative relationship in both locations). Run as the first
+# step of the in-container build command.
+override SONIC_WT_GITDIR_FIXUP := echo "gitdir: ../$(SONIC_WT_PARENT_BASENAME)/.git/worktrees/$(SONIC_WT_BASENAME)" > /$(SONIC_WT_BASENAME)/.git;
+endif
+endif
+endif
+endif
+
 # Consider removing the --ulimit flag once nothing older
 # than Bullseye is being used as a slave container.
 # Build container memory limit (set SONIC_BUILD_MEMORY=none to disable).
@@ -347,6 +409,7 @@ endif
 
 DOCKER_RUN := docker run --rm=true $(DOCKER_PRIVMODE) --init \
     -v $(DOCKER_BUILDER_MOUNT) \
+    $(GIT_WORKTREE_MOUNTS) \
     -v "$(DOCKER_LOCKDIR):$(DOCKER_LOCKDIR)" \
     -w $(DOCKER_BUILDER_WORKDIR) \
     $(SONIC_BUILD_MEMORY_FLAGS) \
@@ -709,7 +772,7 @@ endif
 
 	$(Q)$(DOCKER_RUN) \
 		$(SLAVE_IMAGE):$(SLAVE_TAG) \
-		bash -c "$(SONIC_BUILD_INSTRUCTION) $@ SONIC_BUILD_TARGET=$@; $(COLLECT_BUILD_VERSION); $(SLAVE_SHELL)"
+		bash -c "$(SONIC_WT_GITDIR_FIXUP) $(SONIC_BUILD_INSTRUCTION) $@ SONIC_BUILD_TARGET=$@; $(COLLECT_BUILD_VERSION); $(SLAVE_SHELL)"
 	$(Q)$(docker-image-cleanup)
 
 docker-cleanup:

